File: comment_speed_meter.py
from __future__ import annotations

from collections import deque
from typing import Deque


class CommentSpeedMeter:
    """コメントの風速(コメント/分)を計測するユーティリティ。"""

    def __init__(self, window_seconds: int = 60) -> None:
        if window_seconds <= 0:
            raise ValueError("window_seconds must be positive")
        self.window_seconds = window_seconds
        self._timestamps: Deque[float] = deque()
        self._stream_started_at: float | None = None
        self._total_count = 0

    def start_stream(self, started_at: float) -> None:
        """配信開始時刻を設定し、計測状態をリセットする。"""
        self._stream_started_at = started_at
        self._timestamps.clear()
        self._total_count = 0

    def reset_stream(self) -> None:
        """配信終了時に計測状態をリセットする。"""
        self._stream_started_at = None
        self._timestamps.clear()
        self._total_count = 0

    def ensure_stream_started(self, started_at: float) -> None:
        """配信開始時刻が未設定の場合のみ設定する。"""
        if self._stream_started_at is None:
            self.start_stream(started_at)

    def _prune(self, current_time: float) -> None:
        cutoff = current_time - self.window_seconds
        while self._timestamps and self._timestamps[0] < cutoff:
            self._timestamps.popleft()

    def record(self, timestamp: float) -> None:
        """コメントの受信時刻を記録する。"""
        if self._stream_started_at is None:
            self._stream_started_at = timestamp
        self._prune(timestamp)
        self._timestamps.append(timestamp)
        self._total_count += 1

    def count(self, current_time: float) -> int:
        """計測窓内のコメント数を返す。"""
        self._prune(current_time)
        return len(self._timestamps)

    def rate_per_minute(self, current_time: float) -> int:
        """コメント/分を返す。"""
        count = self.count(current_time)
        if count == 0:
            return 0
        return int(count * 60 / self.window_seconds)

    def total_count(self) -> int:
        """配信開始からのコメント総数を返す。"""
        return self._total_count

    def total_rate_per_minute(self, current_time: float) -> int:
        """配信開始からの平均コメント/分を返す。"""
        if self._stream_started_at is None:
            return 0
        elapsed = current_time - self._stream_started_at
        if elapsed <= 0:
            return 0
        return int(self._total_count * 60 / elapsed)

File: src/chat/comment-speed-meter.ts
/**
 * コメント速度計測（スライディングウィンドウ）
 */
export class CommentSpeedMeter {
  private readonly windowSeconds: number;
  private timestamps: number[] = [];
  private timestampHead = 0;
  private _totalCount = 0;
  private _streamStartedAt: number | null = null;

  constructor(windowSeconds = 60) {
    this.windowSeconds = windowSeconds;
  }

  setState(streamStartedAt: number, totalCount: number): void {
    this._streamStartedAt = streamStartedAt || null;
    this._totalCount = totalCount;
  }

  startStream(timestamp: number): void {
    this._streamStartedAt = timestamp;
    this._totalCount = 0;
    this.timestamps = [];
    this.timestampHead = 0;
  }

  resetStream(): void {
    this._streamStartedAt = null;
    this._totalCount = 0;
    this.timestamps = [];
    this.timestampHead = 0;
  }

  ensureStreamStarted(timestamp: number): void {
    if (this._streamStartedAt === null) {
      this._streamStartedAt = timestamp;
    }
  }

  record(timestamp: number): void {
    this.timestamps.push(timestamp);
    this._totalCount++;
    this._prune(timestamp);
  }

  count(currentTime: number): number {
    this._prune(currentTime);
    return this.timestamps.length - this.timestampHead;
  }

  ratePerMinute(currentTime: number): number {
    const c = this.count(currentTime);
    return Math.round((c / this.windowSeconds) * 60 * 10) / 10;
  }

  totalCount(): number {
    return this._totalCount;
  }

  totalRatePerMinute(currentTime: number): number {
    if (!this._streamStartedAt || this._totalCount === 0) return 0;
    const elapsed = currentTime - this._streamStartedAt;
    if (elapsed <= 0) return 0;
    return Math.round((this._totalCount / elapsed) * 60 * 10) / 10;
  }

  streamStartedAt(): number | null {
    return this._streamStartedAt;
  }

  private _prune(currentTime: number): void {
    const cutoff = currentTime - this.windowSeconds;
    while (
      this.timestampHead < this.timestamps.length &&
      this.timestamps[this.timestampHead] < cutoff
    ) {
      this.timestampHead += 1;
    }

    if (this.timestampHead === this.timestamps.length) {
      this.timestamps = [];
      this.timestampHead = 0;
      return;
    }
    if (
      this.timestampHead >= 1_024 &&
      this.timestampHead * 2 >= this.timestamps.length
    ) {
      this.timestamps = this.timestamps.slice(this.timestampHead);
      this.timestampHead = 0;
    }
  }
}
